avm/ptn/network/hub-networking/modules/subnets.bicep

```diff
@@ -0,0 +1,171 @@
+
+@description('Optional. Enable or disable apply network policies on private link service in the subnet.')
+@allowed([
+  'Disabled'
+  'Enabled'
+  ''
+])
+param privateLinkServiceNetworkPolicies string = ''
+
+@description('Optional. List of address prefixes for the subnet.')
+param addressPrefixes array = []
+
+@description('Optional. Application gateway IP configurations of virtual network resource.')
+param applicationGatewayIPConfigurations array = []
+
+@description('Optional. Array of IpAllocation which reference this subnet.')
+param ipAllocations array = []
+
+@description('Optional. An array of service endpoint policies.')
+param serviceEndpointPolicies array = []
+
+@description('Optional. Array of role assignments to create.')
+param roleAssignments roleAssignmentType
+
+var builtInRoleNames = {
+  Contributor: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c')
+  'Network Contributor': subscriptionResourceId(
+    'Microsoft.Authorization/roleDefinitions',
+    '4d97b98b-1d4f-4787-a291-c67834d212e7'
+  )
+  Owner: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
+  Reader: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId(
+    'Microsoft.Authorization/roleDefinitions',
+    'f58310d9-a9f6-439a-9e8d-f62e7b41a168'
+  )
+  'User Access Administrator': subscriptionResourceId(
+    'Microsoft.Authorization/roleDefinitions',
+    '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9'
+  )
+}
+
+resource virtualNetwork 'Microsoft.Network/virtualNetworks@2023-11-01' existing = {
+  name: virtualNetworkName
+}
+
+resource subnet 'Microsoft.Network/virtualNetworks/subnets@2023-11-01' = {
+  name: name
+  parent: virtualNetwork
+  properties: {
+    addressPrefix: addressPrefix
+    networkSecurityGroup: !empty(networkSecurityGroupResourceId)
+      ? {
+          id: networkSecurityGroupResourceId
+        }
+      : null
+    routeTable: !empty(routeTableResourceId)
+      ? {
+          id: routeTableResourceId
+        }
+      : null
+    natGateway: !empty(natGatewayResourceId)
+      ? {
+          id: natGatewayResourceId
+        }
+      : null
+    serviceEndpoints: serviceEndpoints
+    delegations: delegations
+    privateEndpointNetworkPolicies: !empty(privateEndpointNetworkPolicies) ? any(privateEndpointNetworkPolicies) : null
+    privateLinkServiceNetworkPolicies: !empty(privateLinkServiceNetworkPolicies)
+      ? any(privateLinkServiceNetworkPolicies)
+      : null
+    addressPrefixes: addressPrefixes
+    applicationGatewayIPConfigurations: applicationGatewayIPConfigurations
+    ipAllocations: ipAllocations
+    serviceEndpointPolicies: serviceEndpointPolicies
+  }
+}
+
+resource subnet_roleAssignments 'Microsoft.Authorization/roleAssignments@2022-04-01' = [
```

Comment:
Please note that the role assignment spec was recently changed with a quite drastic impact on the name (and a few other parts). Please update it here and for any other role assignment deployment + udt

Reply:
Almost. Added 2 suggestions for the UDTs. Will resolve this comment though